rag/utils/gaussdb_conn.py

```
@@ -0,0 +1,1431 @@
+            row.setdefault(column, default)
+        return row
+
+    def _build_chunk_upsert(self, index_name: str, rows: list[dict]) -> tuple[str, list[list[Any]]]:
+        if not rows:
+            raise ValueError("rows are required")
+        columns = ordered_columns(rows)
+        table = self.ddl.qualified_name(index_name)
+        placeholders = ", ".join(self._placeholder(column) for column in columns)
+        update_columns = [column for column in columns if column not in {"id", "kb_id"}]
+        update_clause = ", ".join(f"{column} = VALUES({column})" for column in update_columns)
+        sql = f"INSERT INTO {table} ({', '.join(columns)}) VALUES ({placeholders}) ON DUPLICATE KEY UPDATE {update_clause}"
```

Ruff (0.16.0)

[error] Possible SQL injection vector through string-based query construction

(S608)